### ISSUE TITLE
autoexpect does not detect changes in .cljc files

### ISSUE THREAD

**Issue by magnars** [jakemcc/lein-autoexpect#16]
Running `lein autoexpect` works fine with `.cljc` files in Clojure 1.7.0-beta3, but any changes are not detected. Touching a `.clj` file will rerun the tests, but the `.cljc` files are not recompiled.

I see you already have a similar issue for `.cljx` files in #10, but I thought this would warrant a separate issue, since I guess the `.cljc` integration with the clojure compiler is a bit better.

If you have a general idea of how to fix this, but not the time to do so, I'd be willing to work on a pull request with a point in the right direction. :)

**Comment by jakemcc** [comment]
@magnars Pretty sure this just requires an upgrade of tools.namespace to version 0.2.10. Had to do the same thing for `lein-test-refresh` and I forgot about autoexpect.

Just released a new autoexpect (version 1.4.3) that does this. Let me know if it fixes the problem you're seeing.

**Comment by magnars** [comment]
Not quite there, unfortunately.

I have a `prose.cljc` and a `prose_test.clj`. After upgrading:

```
Retrieving lein-autoexpect/lein-autoexpect/1.4.3/lein-autoexpect-1.4.3.jar from clojars
Retrieving org/clojure/tools.namespace/0.2.10/tools.namespace-0.2.10.jar from central
```

I have this behavior:

- touching `prose.cljc` now re-runs the tests, where before nothing happened.
- changes to `prose.cljc` are not visibile, eg. I'm breaking the code, and the tests still pass
- touching `prose_test.clj` gives me this:

    ```
    Error refreshing environment: java.io.FileNotFoundException: Could not locate link_gatekeeper/prose__init.class or link_gatekeeper/prose.clj on classpath: , compiling:(link_gatekeeper/prose_test.clj:1:1)
    ```

The error makes sense in that there is no `prose.clj`, there is only `prose.cljc`.

Thanks so much for looking into this!

**Comment by magnars** [comment]
Another point:

- touching anything after triggering the `Error refreshing environment` results in the same error, and the process must be restarted to get it back in a working state.

**Comment by magnars** [comment]
To be fair, there are quite a few confusing dependencies wrt `tools.namespace` in my project. I'll add a bunch of exclusions and see if that fixes anything.

**Comment by jakemcc** [comment]
`tools.namespace` might not fully support reloading cljc files yet. Looks like there are some [open issues](http://dev.clojure.org/jira/secure/IssueNavigator.jspa?reset=true&jqlQuery=project+%3D+TNS+AND+resolution+%3D+Unresolved+ORDER+BY+priority+DESC%2C+key+DESC&mode=hide) around it.

**Comment by magnars** [comment]
Please disregard the earlier messages. I messed up my repo while switching branches. >.<  Sorry about that.

Let me start over:

------

Not quite there, unfortunately.

I have a `prose.cljc` and a `prose_test.clj`. After upgrading:

```
Retrieving lein-autoexpect/lein-autoexpect/1.4.3/lein-autoexpect-1.4.3.jar from clojars
Retrieving org/clojure/tools.namespace/0.2.10/tools.namespace-0.2.10.jar from central
```

I have this behavior:

- touching `prose.cljc` now re-runs the tests, where before nothing happened.
- changes to `prose.cljc` are not visibile, eg. I'm breaking the code, and the tests still pass
- touching `prose_test.clj` also re-runs the tests.
- changes to `prose_test.clj` are visible, as expected.

**Comment by magnars** [comment]
> tools.namespace might not fully support reloading cljc files yet. Looks like there are some open issues around it.

That would explain it. Thanks.

**Comment by magnars** [comment]
More findings:

I renamed `prose_test.clj` to `prose_test.cljc` and restarted.

Like before:

- changes to `prose.cljc` are not visibile, eg. I'm breaking the code, and the tests still pass.

But check this out:

- changes to `prose_test.cljc` are visible!

Color me particularly confused.

**Comment by magnars** [comment]
And that particular confusion is due to http://dev.clojure.org/jira/browse/TNS-34, "File changes not picked up if reader conditionals used on ns requires".

So that explains it.

**Comment by magnars** [comment]
Yep, as soon as I removed the reader conditionals from the namespace, everything worked wonderfully. Sorry about the message spam.

**Comment by magnars** [comment]
I switched to this tools.namespace fork https://github.com/lorddoig/tools.namespace for a 1.7.0 fix.

**Comment by jakemcc** [comment]
Hmm, good to know there is something out there that works. Hopefully the official `tools.namespace` gets patched.

**Comment by jakemcc** [comment]
@magnars Released version 1.6.0 of lein-autoexpect that depends on the just released tools.namespace 0.2.11 that fixes that issue. Should fix the issue.

**Comment by magnars** [comment]
It does indeed. Thanks! :-)
man. 22. jun. 2015 kl. 01.03 skrev Jake McCrary <notifications@github.com>:
